architecture-decision-records/0013-file-check-queues-and-lambdas.md

```diff
@@ -0,0 +1,185 @@
+rather than hours.
+
+We may add more file checks in future, such as additional checksum algorithms or
+antivirus software. The architecture should be flexible enough to add or remove
+checks later.
+
+[Antivirus scan]: https://github.com/nationalarchives/tdr-antivirus/
+[SHA-256 checksum]: https://github.com/nationalarchives/tdr-checksum/
+[File format ID]: https://github.com/nationalarchives/tdr-file-format/
+
+## Decision
+
+Use AWS [Lambda] to run the file checks, and use [SQS] to coordinate the steps.
+
+See the [architecture diagram] for how the file check workflow fits into the
+overall architecture. The steps in the workflow are:
+
+- A user uploads a file to S3 using the TDR frontend
+- This triggers an upload event, which AWS sends to an SNS topic. The message
+  contains the object's S3 key, which includes the TDR file ID UUID
+- The message is passed to the first SQS queue
+- A Lambda connected to the SQS queue takes the message, extracts the file ID,
+  and [downloads the file][tdr-file-download] from S3 to an [EFS] temporary file
+  store
+- When that Lambda is finished, it adds a message to three further SQS queues,
+  one for each of the backend checks
+- Each SQS message triggers the respective backend check Lambda (checksum, file
+  format ID and antivirus)
+- Each Lambda sends a message to the API update SQS queue containing the results
+  of the check
+- A final Lambda picks up that message, and [sends the results to the consignment
+  API][tdr-api-update], where they are stored in the database
+
+Each check is a single, short process, which makes it very suitable for a
+serverless workflow. If a user uploads thousands of files, we can process them
+in parallel and let SQS and Lambda work through the backlog without having to
+scale any EC2 instances. We are also trying to use serverless services on TDR
+where possible because it removes the need to patch servers.
+
+In all the testing so far, we haven't reached the maximum Lambda execution time
+of 15 minutes. If we do hit those limits, we could consider moving the slow step
+to an ECS task, which doesn't have an execution time limit but is slower to
+start up. We would need some way to trigger the ECS task from SQS, but it would
+otherwise fit into the workflow in the same way that Lambda does at the moment.
+
+We considered using [step functions] instead of SQS to coordinate the tasks. We
+decided against them because when we tried [using them in the Alpha
+prototype][alpha-step-functions], we found that they failed when we hit the
+Lambda concurrent usage limit. This is currently 1000 functions, and we can hit
+this limit if a user uploads a consignment of thousands of small files. We don't
+have this problem with SQS, because the messages wait in the queues until AWS
+lets us start another Lambda.
+
+The initial step of downloading the file from S3 to Lambda is necessary for the
+antivirus and file format ID tasks, which need to access the file as if it was
+on disk.
```

Comment:
This is true but maybe put something to say that we're downloading to EFS rather than using the lambda local disk storage because the lambda disk storage is limited to 512Mb and EFS is effectively unlimited.
I'm not sure if this belongs in this section but it might be worth putting in something to say that the downside of EFS is that as its a network drive, disk access is much slower than using lambda local storage.

Also, as we're only downloading the files once, it speeds up the other lambdas as they don't have to download the same file individually, it simplifies the code a bit and it saves us on data transfer costs.

Reply:
Those are good points, thanks. I've added them. 👍